cpp/modules/deck.gl/core/src/lib/deck.cpp

```diff
@@ -115,12 +115,13 @@ void Deck::run() {
   // TODO(ilija@unfolded.ai): We've got a retain cycle here, revisit
   this->animationLoop->run([&](wgpu::RenderPassEncoder pass) {
     for (auto const& viewport : this->viewManager->getViewports()) {
+      // Expose the current viewport to layers for project* function
+      this->layerManager->activateViewport(viewport);
+
       for (auto const& layer : this->layerManager->layers) {
         // TODO(ilija@unfolded.ai): Pass relevant layer properties to getUniformsFromViewport
-        auto viewportUniforms = getUniformsFromViewport(viewport);
-
-        // TODO(ilija@unfolded.ai): Remove and use row_major specifier in uniform layout once drawing works
-        viewportUniforms.viewProjectionMatrix = viewportUniforms.viewProjectionMatrix.Transpose();
+        // TODO(ilija@unfolded.ai): Find a way to detect device pixel ratio and pass it here
```

Comment:
I'd try searching Metal documentation for details on this.

So our drawing buffer is set up to be 2x the width/height? Is that done automatically by the swap chain?

The mapping of drawing buffer to screen is controlled by `gl.viewport` in WebGL, are there counterparts in WebGPU?

Reply:
It looks like scaling wasn't taken into account in the Metal binding. GLSL takes care of updating this, but since the swapchain implementation declares a new layer, it has to be accounted for as part of the drawable size.

Added that as well as plumbing for retrieving it out of glfw through our animation loop